File: examples/pylva-langgraph-js-cost-tracking-starter/basic_graph.ts
import { Annotation, END, START, StateGraph } from '@langchain/langgraph';
import { ChatOpenAI } from '@langchain/openai';
import { PylvaCallbackHandler } from '@pylva/sdk/langgraph';

const GraphState = Annotation.Root({
  question: Annotation<string>(),
  answer: Annotation<string>(),
});

const model = new ChatOpenAI({
  model: 'gpt-4o-mini',
});

const graph = new StateGraph(GraphState)
  .addNode('answer_question', async (state) => {
    const response = await model.invoke([
      {
        role: 'user',
        content: state.question,
      },
    ]);

    return { answer: String(response.content) };
  })
  .addEdge(START, 'answer_question')
  .addEdge('answer_question', END)
  .compile();

const handler = new PylvaCallbackHandler({
  apiKey: process.env.PYLVA_API_KEY!,
});

const result = await graph.invoke(
  { question: 'Give me one practical way to reduce LLM cost.' },
  {
    callbacks: [handler],
    metadata: { pylva_customer_id: 'cust_acme' },
  },
);

console.log(result.answer);

File: examples/pylva-langgraph-cost-tracking-starter/basic_graph.py
from __future__ import annotations

import os
from typing_extensions import TypedDict

from langchain.chat_models import init_chat_model
from langchain_core.messages import HumanMessage
from langgraph.graph import END, START, StateGraph
from pylva.langchain import PylvaCallbackHandler


class State(TypedDict):
    question: str
    answer: str


model = init_chat_model("openai:gpt-4o-mini")


def answer(state: State) -> dict[str, str]:
    response = model.invoke([HumanMessage(content=state["question"])])
    return {"answer": str(response.content)}


graph = (
    StateGraph(State)
    .add_node("answer", answer)
    .add_edge(START, "answer")
    .add_edge("answer", END)
    .compile()
)

handler = PylvaCallbackHandler(api_key=os.environ["PYLVA_API_KEY"])

result = graph.invoke(
    {"question": "Give me a one sentence explanation of marginal cost."},
    config={
        "callbacks": [handler],
        "metadata": {"pylva_customer_id": "cust_demo"},
    },
)

print(result["answer"])
